sync-android/src/main/java/com/cloudant/sync/datastore/encryption/KeyManager.java

```diff
@@ -0,0 +1,217 @@
+        } catch (Exception e) {
+            throw new DPKException("Failed to decrypt DPK", e);
+        }
+        return dpk;
+    }
+
+    /**
+     * Generates a Data Protection Key (DPK), encrypts it, and stores it inside the {@link
+     * SharedPreferences}.
+     *
+     * @param password Password used to encrypt the DPK
+     * @return The DPK
+     */
+    public EncryptionKey generateAndSaveKeyProtectedByPassword(String password) {
+        EncryptionKey dpk = null;
+        if (password == null || password.equals("")) {
+            throw new IllegalArgumentException("password is required to be a non-null/non-empty " +
+                    "string");
+        }
+        try {
+            if (!keyExists()) {
+                byte[] dpkBytes = generateSecureRandomBytesWithLength
+                        (CDTENCRYPTION_KEYCHAIN_ENCRYPTIONKEY_SIZE);
+                byte[] salt = generateSecureRandomBytesWithLength
+                        (CDTENCRYPTION_KEYCHAIN_PBKDF2_SALT_SIZE);
+                byte[] iv = generateSecureRandomBytesWithLength
+                        (CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE);
+                SecretKey aesKey = pbkdf2DerivedKeyForPassword(password, salt,
+                        CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS,
+                        CDTENCRYPTION_KEYCHAIN_AES_KEY_SIZE);
+
+                byte[] encryptedDpkBytes = DKPEncryptionUtil.encryptAES(aesKey, iv, dpkBytes);
+
+                KeyData keyData = new KeyData(encryptedDpkBytes, salt, iv,
+                        CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS, CDTENCRYPTION_KEYCHAIN_VERSION);
+
+                if (this.storage.saveEncryptionKeyData(keyData)) {
+                    dpk = new EncryptionKey(dpkBytes);
+                }
+            }
+        } catch (Exception e) {
+            throw new DPKException("Failed to encrypt DPK", e);
+        }
+        return dpk;
+    }
+
+    /**
+     * Checks if the encrypted Data Protection Key (DPK) is inside the {@link SharedPreferences}.
+     *
+     * @return true if the encrypted DPK is inside the {@link SharedPreferences}, false otherwise
+     */
+    public boolean keyExists() {
+        return this.storage.encryptionKeyDataExists();
+    }
+
+    /**
+     * Clears security metadata from the {@link SharedPreferences}.
+     *
+     * @return Success (true) or failure (false)
+     */
+    public boolean clearKey() {
+        return this.storage.clearEncryptionKeyData();
+    }
+
+    // PRIVATE HELPER METHODS
+    private boolean validateEncryptionKeyData(KeyData data) {
+        if (data.getIv().length != CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE) {
+            LOGGER.warning("IV does not have the expected size: " +
+                    CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE + " bytes");
+            return false;
+        }
+        return true;
+    }
+
+    private SecretKey pbkdf2DerivedKeyForPassword(String password, byte[] salt, int
```

Comment:
I just moved them internal because they were not actually used by other code.  I felt like inclusion in a Util would be justified by reuse.  I am open to moving them out if there is a strong preference.